plugins/nordic-hid/fu-nordic-hid-archive.c

```diff
@@ -11,15 +11,161 @@
 #include "fu-nordic-hid-firmware-b0.h"
 #include "fu-nordic-hid-firmware-mcuboot.h"
 
-/* current version format is 0 */
-#define MAX_VERSION_FORMAT 0
+/* the plugin currently supports version format of either 0 or 1 */
+#define MIN_VERSION_FORMAT 0
+#define MAX_VERSION_FORMAT 1
 
 struct _FuNordicHidArchive {
 	FuFirmwareClass parent_instance;
 };
 
 G_DEFINE_TYPE(FuNordicHidArchive, fu_nordic_hid_archive, FU_TYPE_FIRMWARE)
 
+static const gchar *
+fu_nordic_hid_archive_parse_file_get_bootloader_name(JsonObject *obj, GError **error)
+{
+	if (json_object_has_member(obj, "version_B0")) {
+		return "B0";
+	} else if (json_object_has_member(obj, "version_MCUBOOT")) {
+		return "MCUBOOT";
+	} else if (json_object_has_member(obj, "version_MCUBOOT+XIP")) {
+		return "MCUBOOT+XIP";
+	} else {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "only B0 and MCUboot bootloaders are supported");
+		return NULL;
+	}
+}
+
+static FuFirmware *
+fu_nordic_hid_archive_parse_file_image_create(const gchar *bootloader_name, GError **error)
+{
+	if (!g_strcmp0(bootloader_name, "B0")) {
+		return g_object_new(FU_TYPE_NORDIC_HID_FIRMWARE_B0, NULL);
+	} else if (!g_strcmp0(bootloader_name, "MCUBOOT") ||
+		   !g_strcmp0(bootloader_name, "MCUBOOT+XIP")) {
+		return g_object_new(FU_TYPE_NORDIC_HID_FIRMWARE_MCUBOOT, NULL);
+	} else {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "only B0 and MCUboot bootloaders are supported");
+		return NULL;
+	}
+}
+
+static gchar *
+fu_nordic_hid_archive_parse_file_get_board_name(JsonObject *obj,
+						gint64 manifest_ver,
+						GError **error)
+{
+	g_auto(GStrv) board_split = NULL;
+	const gchar *board_name_readout = NULL;
+
+	if (!json_object_has_member(obj, "board")) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "manifest invalid as has no target board information");
+		return NULL;
+	}
+
+	board_name_readout = json_object_get_string_member(obj, "board");
+	if (manifest_ver == 0) {
+		/* for manifest "format-version" of "0", the board name is represented only
+		 * by part of the string before the "_" symbol
+		 */
+		board_split = g_strsplit(board_name_readout, "_", -1);
+		if (board_split[0] == NULL) {
+			g_set_error_literal(error,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INVALID_FILE,
+					    "manifest invalid as has no board information");
+			return NULL;
+		}
+		return g_strdup(board_split[0]);
+	} else {
+		/* duplicate string for consistent memory management */
+		return g_strdup(board_name_readout);
+	}
+}
+
+static gint
+fu_nordic_hid_archive_parse_file_get_flash_area_id(JsonObject *obj,
+						   gint64 manifest_ver,
+						   guint file_idx,
+						   const gchar *bootloader_name,
+						   guint files_cnt,
+						   GError **error)
+{
+	if (manifest_ver == 0) {
+		/* the images are expected to be listed in strict order */
+		return file_idx;
+	} else {
```

Comment:
you don;t need this else section -- and coverity will hate it :)

Reply:
I wanted to explicitly indicate that the branches here are alternatives to each other. I will use a temporary value without an early return then.